### ISSUE TITLE
Incorrectly reported borrow error instead of bad signature when using tuples

### ISSUE THREAD

**Issue by HollayHorvath** [rust-lang/rust#107363]
The following code
```rust
fn a(_: &State) {}
fn b(_: &State) {}

struct State {}

fn expected_bad_signature_error() {
    let fns: Vec<fn(_)> = vec![a, b];
    
    for test in fns {
        let state = State {};
        
        test(&state);
    }
}
```
 reports a bad signature error as expected:
```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
 --> src/lib.rs:7:35
  |
7 |     let fns: Vec<fn(_)> = vec![a, b];
  |                                   ^ expected fn pointer, found fn item
  |
  = note: expected fn pointer `fn(&State)`
                found fn item `for<'a> fn(&'a State) {b}`
```

but changing from a single value to tuples
```rust
fn unexpected_borrow_error() {
    let fns: Vec<(fn(_),)> = vec![
        (a,),
        (b,),
    ];
    
    for (test,) in fns {
        let state = State {};
        
        test(&state);
    }
}
```
 causes the compiler to report a borrow error:
```
error[[E0597]](https://doc.rust-lang.org/stable/error-index.html#E0597): `state` does not live long enough
  --> src/lib.rs:25:14
   |
22 |     for (test,) in fns {
   |                    --- borrow later used here
...
25 |         test(&state);
   |              ^^^^^^ borrowed value does not live long enough
26 |     }
   |     - `state` dropped here while still borrowed
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dd05e32eae9023542dba4e84a97e4384) 

To clarify, my issue is not the error itself (the correct type is `fn(&_)`), but the incorrect error being reported for the `unexpected_borrow_error` case.

@vincenzopalazzo helped me double-check this issue
